lib/LedgerSMB/Database/Change.pm

```diff
@@ -178,47 +184,38 @@ Applies the current file to the db in the current dbh.
 
 sub apply {
     my ($self, $dbh) = @_;
-    my $need_commit = _need_commit($dbh);
     my $before = '';
-    my $after;
     my $sha = $dbh->quote($self->sha);
     my $path = $dbh->quote($self->path);
-    my $no_transactions = $self->{properties}->{no_transactions};
-    if ($self->is_applied($dbh)){
-        $after = "
-              UPDATE db_patches
-                     SET last_updated = now()
-               WHERE sha = $sha;
-        ";
-    } else {
-        $after = "
-           INSERT INTO db_patches (sha, path, last_updated)
-           VALUES ($sha, $path, now());
-        ";
-    }
+    my $no_transactions = !!$self->{properties}->{no_transactions};
+    $dbh->{AutoCommit} = !$no_transactions;
+
+    my $after = $self->is_applied($dbh)
+              ? "
+                  UPDATE db_patches
+                         SET last_updated = now()
+                   WHERE sha = $sha;
+              "
+              : "
+                   INSERT INTO db_patches (sha, path, last_updated)
+                   VALUES ($sha, $path, now());
+                ";
+    my $success = $dbh->prepare($self->content_wrapped($before, $after))->execute();
+    die "$DBI::state: $DBI::errstr while applying $path"
+        unless $success or $no_transactions;
+    $dbh->commit unless !$success || !$no_transactions;
```

Comment:
why are you `die`-ing when the $no_transactions condition is true? or does the above mean `(die ... unless) or $no_transactions`? If the latter, that's too obfuscated for me. If not, apparently, we always need to die when `$no_transactions` is true?

Reply:
If we applied the patch or if we are in a transaction, don't die. The former is we succeeded, the latter is it will be rolled back later.